ath79: support for TP-Link EAP225-Outdoor v3

yksflip opened the pull request:
This Commit adds support for the TP-Link EAP225-Outdoor v3.  This model is almost identical to the EAP225-Outdoor v1. The major difference is the RTL8211fs PHY Chipset.

There might also changes required in  `firmware-utils/src/tplink-safeloader.c`. I drafted them [here](https://github.com/yksflip/firmware-utils/commit/aa30488d9d2ce2a961193e659f07224e99c606b6), but I'm unsure where to send a patch for that, as it's an separate package now?

See also the related [forum thread](https://forum.openwrt.org/t/support-for-tp-link-eap225-outdoor-v3/123132/4)

Signed-off-by: Philipp Rothmann <philipprothmann@posteo.de>

cuechan committed:
ath79: support for TP-Link EAP225-Outdoor v3

This model is almost identical to the EAP225-Outdoor v1.
Major difference is the RTL8211FS PHY Chipset.

Device specifications:
* SoC: QCA9563 @ 775MHz
* Memory: 128MiB DDR2
* Flash: 16MiB SPI-NOR
* Wireless 2.4GHz (SoC): b/g/n 2x2
* Wireless 5GHz (QCA9886): a/n/ac 2x2 MU-MIMO
* Ethernet (RTL8211FS): 1× 1GbE, PoE

Flashing instructions:
* ssh into target device with recent (>= v1.6.0) firmware
* run `cliclientd stopcs` on target device
* upload factory image via web interface

Debricking:
To recover the device, you need access to the serial port. This requires
fine soldering to test points, or the use of probe pins.
* Open the case and solder wires to the test points: RXD, TXD and TPGND4
  * Use a 3.3V UART, 115200 baud, 8n1
* Interrupt bootloader by holding ctrl+B during boot
* upload initramfs via built-in tftp client and perform sysupgrade
    setenv ipaddr 192.168.1.1 # default, change as required
    setenv serverip 192.168.1.10 # default, change as required
    tftp 0x80800000 initramfs.bin
    bootelf $fileaddr

MAC addresses:
MAC address (as on device label) is stored in device info partition at
an offset of 8 bytes. ath9k device has same address as ethernet, ath10k
uses address incremented by 1.
From stock ifconfig:

    ath0      Link encap:Ethernet  HWaddr D8:...:2E
    ath10     Link encap:Ethernet  HWaddr D8:...:2F
    br0       Link encap:Ethernet  HWaddr D8:...:2E
    eth0      Link encap:Ethernet  HWaddr D8:...:2E

Signed-off-by: Paul Maruhn <paulmaruhn@posteo.de>
Co-developed-by: Philipp Rothmann <philipprothmann@posteo.de>
Signed-off-by: Philipp Rothmann <philipprothmann@posteo.de>
[Add pre-calibraton nvme-cells]
Tested-by: Tido Klaassen <tido_ff@4gh.eu>
Signed-off-by: Nick Hainke <vincent@systemli.org>

yksflip commented:
the signed-off's should hopefully be correct now ...

jp-bennett reviewed (approved):
For what it's worth, Reviewed-by: Jonathan Bennett <jbennett@incomsystems.biz>
I've used these changes to also make the EAP-225 V4 work. Once this lands, I'll launch the PR for that board, too.

fogs commented:
Would love to see this merged; the device is great. What needs to be done here?

jp-bennett commented:
@fogs This device and it's near clone, the EAP-225 v4 may have a potential issue where the Ethernet phy doesn't properly reset on reboot, meaning that uboot can't use the Ethernet on a reboot. It has to get a full power cycle for uboot to see the network.  See https://forum.openwrt.org/t/support-for-tp-link-eap225-outdoor-v3/123132/27

PolynomialDivision commented:
I would say your commit message it to short. Just put the information about hardware capabilities and how to install again into the commit description, although it might be identical to the v1 version.

PolynomialDivision commented:
Ping @svanheule

svanheule commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -401,6 +401,17 @@ define Device/tplink_eap225-outdoor-v1
+define Device/tplink_eap225-outdoor-v3
+  $(Device/tplink-eap2x5)
+  SOC := qca9563
+  IMAGE_SIZE := 13824k
```
This is the image size for the partition layout from eap2x5-1port.dtsi. Here you overwrite the partition layout, so you *could* `IMAGE_SIZE := 15104k`. That's 1280kB, but shouldn't be an issue for now. I think we should either keep the original layout, or change it for all devices (this was discussed previously, but I would have to check some old PRs and the mailing list).

jwmullally replied:
Unless there is some showstopping reason, as a general policy IMAGE_SIZE, DTS "firmware" and tplink-safeloader "firmware" should always be the same size and only overwrite `os-image` and `file-system`.

When they are different it just causes too much of a mess, is confusing, unnecessarily introduces edge cases, no-one can remember why it was done like that, it breaks sysupgrade-to-OEM (even if TFTP is still the main way to restore), etc. People end up just blindly copy-pasting the pattern because its too confusing to trace and its whats the older devices do, instead of matching the current "partition-table" in the OEM firmware (E.g. the CPE devices).

Much simpler and safer just to keep them all the same as intended, which thankfully seems to be the case for this patch:) Just commenting for future reference!

svanheule reviewed (changes requested):
You should add more device info, install information, debricking instructions, original MAC address offsets (did I miss something?) to the commit message. See other (TP-Link EAP) devices for some examples.

I'm also preparing some changes to refactor the DTS files for the EAP series a bit: https://git.openwrt.org/?p=openwrt/staging/svanheule.git;a=shortlog;h=refs/heads/patch/eap-nvmem-caldata

Let's see how that goes, but my comments are with these pending changes in mind.

svanheule commented:
I've just pushed some changes for the EAP2xx DTS files, so these patches will need to be rebased. The current DTS file for EAP225 v3 and commit eca0d7301198febc64f4a3561d50dac99d931155 should also help to show how you can use `nvmem-cell` for the wireless calibration data.

yksflip commented:
thanks for your comments! I tried to resolve your them as good as I can.
I rebased it to `openwrt/staging/svanheule.git` and could build it without problems.

Regarding the reset-gpio, is it correct to just leave it out (as the discussion showed that it's probably not connected)?

As I'm traveling, I unfortunately can't test the changes on the device at the moment. But I asked someone to test it and verify the mac address offsets, as I'm not sure about it and can't find it in my notes anymore.

svanheule commented:
I have been made aware earlier today that 11ac Wave-2 hardware requires a "pre-calibration" nvmem-cell, not "calibration". See for example https://git.openwrt.org/?p=openwrt/staging/svanheule.git;a=commitdiff;h=5ad174c860ee8add2c7ee1a80d528377d452dbb5

PolynomialDivision commented:
> I have been made aware earlier today that 11ac Wave-2 hardware requires a "pre-calibration" nvmem-cell, not "calibration". See for example https://git.openwrt.org/?p=openwrt/staging/svanheule.git;a=commitdiff;h=5ad174c860ee8add2c7ee1a80d528377d452dbb5

Heyho, I added the rest, to get this upstream:
https://github.com/PolynomialDivision/openwrt/commit/70a8016541b442d83345bea0fe23cae5cc7a1195
Also found someone who tested the changes:
```
[   40.536057] ath10k 5.15 driver, optimized for CT firmware, probing pci device: 0x56.
[   40.545850] ath10k_pci 0000:00:00.0: enabling device (0000 -> 0002)
[   40.552675] ath10k_pci 0000:00:00.0: pci irq legacy oper_irq_mode 1 irq_mode 0 reset_mode 0
[   43.795922] ath10k_pci 0000:00:00.0: qca9888 hw2.0 target 0x01000000 chip_id 0x00000000 sub 0000:0000
[   43.805486] ath10k_pci 0000:00:00.0: kconfig debug 0 debugfs 1 tracing 0 dfs 1 testmode 0
[   43.824762] ath10k_pci 0000:00:00.0: firmware ver 10.4b-ct-9888-fW-13-5ae337bb1 api 5 features mfp,peer-flow-ctrl,txstatus-noack,wmi-10.x-CT,ratemask-CT,regdump-CT,txrate-CT,flush-all-CT,pingpong-CT,ch-regs-CT,nop-CT,set-special-CT,tx-rc-CT,cust-stats-CT,txrate2-CT,beacon-cb-CT,wmi-block-ack-CT,wmi-bcn-rc-CT crc32 59e741e7
[   44.167546] ath10k_pci 0000:00:00.0: Loading BDF type 0
[   44.178368] ath10k_pci 0000:00:00.0: board_file api 2 bmi_id 0:24 crc32 465094d5
[   45.948770] ath10k_pci 0000:00:00.0: 10.4 wmi init: vdevs: 16  peers: 48  tid: 96
[   45.956562] ath10k_pci 0000:00:00.0: msdu-desc: 2500  skid: 32
[   46.006509] ath10k_pci 0000:00:00.0: wmi print 'P 48/48 V 16 K 144 PH 176 T 186  msdu-desc: 2500  sw-crypt: 0 ct-sta: 0'
[   46.017817] ath10k_pci 0000:00:00.0: wmi print 'free: 114572 iram: 12644 sram: 29508'
[   46.253922] ath10k_pci 0000:00:00.0: htt-ver 2.2 wmi-op 6 htt-op 4 cal pre-cal-nvmem max-sta 32 raw 0 hwcrypto 1
```

svanheule commented:
Thanks, @PolynomialDivision ! I will look into merging support later this week. Enabling CONFIG_REALTEK_PHY should be separate patch IMHO, but I can do this myself if needed. That way the change that affects the entire ath79-generic subtarget is separate from the device patch.

yksflip committed:
ath79: generic: add support for Realtek PHY

Some models of the TP-Link EAP225 series use a Realtek PHY,
therefore the driver is added.

Signed-off-by: Philipp Rothmann <philipprothmann@posteo.de>

yksflip commented:
hope that's alright :)

svanheule commented:
Nearly there! ;)

The Realtek phy config commit should come first, otherwise your device won't work properly until after that second commit. Unless it works without that driver, but it improves the device support. Maybe then it can come after the device, but the commit message would have to argue why these APs need that driver. Also, I would prefer it if you fix the capitalization of "TP-Link EAP225". And 'thatfore' should probably be 'therefore'.

PolynomialDivision commented:
Thanks! :D
